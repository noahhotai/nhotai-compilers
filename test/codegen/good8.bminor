add_one1: function integer(i: integer, x:string) = {
    
    return 1 + i;
}

add_one: function integer(i: integer, x:string) = {

    return 1 + i + add_one1(i, x);
}

main: function integer () = {
    // testing several function call and return value
    print add_one(1, "hello\n");
}
