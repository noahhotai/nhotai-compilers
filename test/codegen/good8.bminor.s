.file "test/codegen/good8.bminor.s"
.text
.global add_one1
add_one1:
PUSHQ %rbp
MOVQ %rsp, %rbp
PUSHQ %rdi
PUSHQ %rsi
PUSHQ %rdx
PUSHQ %rcx
PUSHQ %r8
PUSHQ %r9
MOVQ $1, %rbx
MOVQ -8(%rbp), %r10
ADDQ %rbx, %r10
MOV %r10, %rax
ADDQ $0, %rsp
POPQ %r9
POPQ %r8
POPQ %rcx
POPQ %rdx
POPQ %rsi
POPQ %rdi
POPQ %rbp
RET
RET
ADDQ $0, %rsp
POPQ %r9
POPQ %r8
POPQ %rcx
POPQ %rdx
POPQ %rsi
POPQ %rdi
POPQ %rbp
RET
.text
.global add_one
add_one:
PUSHQ %rbp
MOVQ %rsp, %rbp
PUSHQ %rdi
PUSHQ %rsi
PUSHQ %rdx
PUSHQ %rcx
PUSHQ %r8
PUSHQ %r9
MOVQ $1, %rbx
MOVQ -8(%rbp), %r10
MOVQ -8(%rbp), %r11
MOVQ %r11, %rdi
MOVQ -16(%rbp), %r11
MOVQ %r11, %rsi
PUSHQ %r10
PUSHQ %r11
PUSHQ %rbx
PUSHQ %r12
PUSHQ %r13
PUSHQ %r14
PUSHQ %r15
CALL add_one1
POPQ %rbx
POPQ %r12
POPQ %r13
POPQ %r14
POPQ %r15
POPQ %r10
POPQ %r11
MOVQ %rax, %r11
ADDQ %r10, %r11
ADDQ %rbx, %r11
MOV %r11, %rax
ADDQ $0, %rsp
POPQ %r9
POPQ %r8
POPQ %rcx
POPQ %rdx
POPQ %rsi
POPQ %rdi
POPQ %rbp
RET
RET
ADDQ $0, %rsp
POPQ %r9
POPQ %r8
POPQ %rcx
POPQ %rdx
POPQ %rsi
POPQ %rdi
POPQ %rbp
RET
.text
.global main
main:
PUSHQ %rbp
MOVQ %rsp, %rbp
PUSHQ %rdi
PUSHQ %rsi
PUSHQ %rdx
PUSHQ %rcx
PUSHQ %r8
PUSHQ %r9
MOVQ $1, %rbx
MOVQ %rbx, %rdi
.data
.L1:
.string "hello\n"
.text
MOVQ $.L1, %rbx
MOVQ %rbx, %rsi
PUSHQ %r10
PUSHQ %r11
PUSHQ %rbx
PUSHQ %r12
PUSHQ %r13
PUSHQ %r14
PUSHQ %r15
CALL add_one
POPQ %rbx
POPQ %r12
POPQ %r13
POPQ %r14
POPQ %r15
POPQ %r10
POPQ %r11
MOVQ %rax, %rbx
MOVQ %rbx, %rdi
CALL print_integer
ADDQ $0, %rsp
POPQ %r9
POPQ %r8
POPQ %rcx
POPQ %rdx
POPQ %rsi
POPQ %rdi
POPQ %rbp
RET
