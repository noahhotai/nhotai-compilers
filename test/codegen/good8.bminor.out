Program parsed successfully.
i resolves to param 0
i resolves to param 0
add_one1 resolves to global 0
i resolves to param 0
x resolves to param 1
add_one resolves to local 0
Program resolved successfully.
Program type checked successfully.
